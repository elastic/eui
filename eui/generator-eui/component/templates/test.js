import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test';

import { <%= componentName %> } from './<%= fileName %>';

describe('<%= componentName %>', () => {
  test('is rendered', () => {
    const component = render(
      <<%= componentName %> {...requiredProps} />
    );

    expect(component)
      .toMatchSnapshot();
  });
});
