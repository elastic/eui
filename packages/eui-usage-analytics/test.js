/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const Codeowners = require('codeowners');
const temp = new Codeowners("../../../kibana");

const { scan } = require('./scan');

const runScan = async () => {
  const scanResult = await scan();
  console.log(scanResult);
};

runScan();
