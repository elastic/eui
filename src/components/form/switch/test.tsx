import React from 'react';

import { EuiSwitch } from './switch';

export const Test = ({ checked }: { checked: boolean }) => (
  <EuiSwitch
    checked={checked}
    label="Test"
    onChange={e => console.log(e.target.checked)}
  />
);
