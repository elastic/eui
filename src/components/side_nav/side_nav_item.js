import React, {
  cloneElement,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import {
  EuiIcon,
} from '../icon';

const defaultRenderItem = ({ href, onClick, className, children, ...rest }) => {
  if (href) {
    return (
      <a
        className={className}
        href={href}
        {...rest}
      >
        {children}
      </a>
    );
  }

  else if (onClick) {
    return (
      <button
        className={className}
        onClick={onClick}
        {...rest}
      >
        {children}
      </button>
    );
  }

  return (
    <div
      className={className}
      {...rest}
    >
      {children}
    </div>
  );
};

export const EuiSideNavItem = ({
  isOpen,
  isSelected,
  isParent,
  icon,
  onClick,
  href,
  items,
  children,
  depth,
  renderItem = defaultRenderItem,
  ...rest,
}) => {
  let childItems;

  if (items && isOpen) {
    childItems = (
      <div className="euiSideNavItem__items">
        {items}
      </div>
    );
  }

  let buttonIcon;

  if (icon) {
    buttonIcon = cloneElement(icon, {
      className: 'euiSideNavItemButton__icon',
    });
  }

  const classes = classNames('euiSideNavItem', {
    'euiSideNavItem--root': depth === 0,
    'euiSideNavItem--rootIcon': depth === 0 && icon,
    'euiSideNavItem--trunk': depth === 1,
    'euiSideNavItem--branch': depth > 1,
    'euiSideNavItem--hasChildItems': !!childItems
  });

  const buttonClasses = classNames('euiSideNavItemButton', {
    'euiSideNavItemButton--isClickable': onClick || href,
    'euiSideNavItemButton-isOpen': depth > 0 && isOpen && !isSelected,
    'euiSideNavItemButton-isSelected': isSelected,
  });

  let caret;

  if (depth > 0 && isParent && !isOpen && !isSelected) {
    caret = <EuiIcon type="arrowDown" color="subdued" size="s" />;
  }

  const buttonContent = (
    <span className="euiSideNavItemButton__content">
      {buttonIcon}

      <span className="euiSideNavItemButton__label">
        {children}
      </span>

      {caret}
    </span>
  );

  return (
    <div className={classes}>
      {renderItem({ href, onClick, className: buttonClasses, children: buttonContent, ...rest })}
      {childItems}
    </div>
  );
};

EuiSideNavItem.propTypes = {
  isOpen: PropTypes.bool,
  isSelected: PropTypes.bool,
  isParent: PropTypes.bool,
  icon: PropTypes.node,
  onClick: PropTypes.func,
  href: PropTypes.string,
  items: PropTypes.node,
  children: PropTypes.node,
  depth: PropTypes.number,
  renderItem: PropTypes.func,
};
