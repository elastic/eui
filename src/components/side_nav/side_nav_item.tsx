/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  cloneElement,
  ReactNode,
  ReactElement,
  MouseEventHandler,
  useState,
  useEffect,
} from 'react';
import classNames from 'classnames';

import { CommonProps } from '../common';

import { EuiIcon } from '../icon';

import { getSecureRelForTarget } from '../../services';
import { validateHref } from '../../services/security/href_validator';
import { EuiInnerText } from '../inner_text';

/**
 * The props that are exposed to, or altered for, the consumer
 * for use in the object of items in `EuiSideNav`
 * can be found in the `side_nave_types.ts` file.
 */

export type _EuiSideNavItemButtonProps = CommonProps & {
  /**
   * Is an optional string to be passed as the navigation item's `href` prop,
   * and by default it will force rendering of the item as an `<a>`
   */
  href?: string;
  target?: string;
  rel?: string;
  /**
   * Callback function to be passed as the navigation item's `onClick` prop,
   * and by default it will force rendering of the item as a `<button>` instead of a link
   */
  onClick?: MouseEventHandler<HTMLButtonElement | HTMLElement>;
  children: ReactNode;
  disabled?: boolean;
};

export interface _EuiSideNavItemProps {
  /**
   * React node which will be rendered as a small icon to the
   * left of the navigation item text
   */
  icon?: ReactElement;
  /**
   * If set to true it will render the item in a visible
   * "selected" state, and will force all ancestor navigation items
   * to render in an "open" state
   */
  isSelected?: boolean;
  /**
   * Enhances the whole item's section (including nested items) with
   * a slight background and bold top item
   */
  emphasize?: boolean;
  /**
   * Restrict the item's text length to a single line
   */
  truncate?: boolean;
  /**
   * Passed to the actual `.euiSideNavItemButton` element
   */
  buttonClassName?: string;
  // Exposed as different prop type to consumer
  items?: ReactNode;
  // Not exposed to consumer
  isOpen?: boolean;
  isParent?: boolean;
  depth?: number;
  childrenOnly?: boolean;
}

type ExcludeEuiSideNavItemProps<T> = Pick<
  T,
  Exclude<keyof T, keyof _EuiSideNavItemProps | 'renderItem'>
>;
type OmitEuiSideNavItemProps<T> = {
  [K in keyof ExcludeEuiSideNavItemProps<T>]: T[K];
};

export type RenderItem<T> = (
  // argument is the set of extra component props + _EuiSideNavItemButtonProps
  props: OmitEuiSideNavItemProps<T> & _EuiSideNavItemButtonProps
) => JSX.Element;

export type EuiSideNavItemProps<T> = T extends { renderItem: Function }
  ? T & { renderItem: RenderItem<T> }
  : T;

const DefaultRenderItem = ({
  href,
  target,
  rel,
  onClick,
  className,
  children,
  disabled,
  ...rest
}: _EuiSideNavItemButtonProps) => {
  if (href && !disabled) {
    const secureRel = getSecureRelForTarget({ href, rel, target });
    return (
      <a
        className={className}
        href={href}
        target={target}
        rel={secureRel}
        onClick={onClick}
        {...rest}>
        {children}
      </a>
    );
  }

  if (onClick || disabled) {
    return (
      <button
        type="button"
        className={className}
        onClick={onClick}
        disabled={disabled}
        {...rest}>
        {children}
      </button>
    );
  }

  return (
    <div className={className} {...rest}>
      {children}
    </div>
  );
};

export function EuiSideNavItem<
  T extends _EuiSideNavItemButtonProps &
    _EuiSideNavItemProps & { renderItem?: (props: any) => JSX.Element }
>({
  isOpen,
  isSelected,
  isParent,
  icon,
  onClick,
  href: _href,
  rel,
  target,
  items,
  children,
  renderItem: RenderItem = DefaultRenderItem,
  depth = 0,
  className,
  truncate = true,
  emphasize,
  buttonClassName,
  childrenOnly,
  ...rest
}: EuiSideNavItemProps<T>) {
  const isHrefValid = !_href || validateHref(_href);
  const href = isHrefValid ? _href : '';
  const isClickable = onClick || href;

  // Forcing accordion style item if not linked, but has children
  const [itemIsOpen, setItemIsOpen] = useState(isOpen);
  useEffect(() => {
    setItemIsOpen(isOpen);
  }, [isOpen]);

  const toggleItemOpen = () => {
    setItemIsOpen((isOpen) => !isOpen);
  };

  let childItems;
  if (items && itemIsOpen) {
    childItems = <div className="euiSideNavItem__items">{items}</div>;
  }

  let buttonIcon;
  if (icon) {
    buttonIcon = cloneElement(icon, {
      className: classNames('euiSideNavItemButton__icon', icon.props.className),
    });
  }

  const classes = classNames(
    'euiSideNavItem',
    {
      'euiSideNavItem--root': depth === 0,
      'euiSideNavItem--rootIcon': depth === 0 && icon,
      'euiSideNavItem--trunk': depth === 1,
      'euiSideNavItem--branch': depth > 1,
      'euiSideNavItem--hasChildItems': !!childItems,
      'euiSideNavItem--emphasized': emphasize,
    },
    className
  );

  const buttonClasses = classNames(
    'euiSideNavItemButton',
    {
      'euiSideNavItemButton--isClickable': isClickable,
      'euiSideNavItemButton-isOpen': depth > 0 && itemIsOpen && !isSelected,
      'euiSideNavItemButton-isSelected': isSelected,
    },
    buttonClassName
  );

  let caret;

  if (depth > 0 && childrenOnly) {
    caret = <EuiIcon type={itemIsOpen ? 'arrowDown' : 'arrowRight'} size="s" />;
  }

  const buttonContent = (
    <span className="euiSideNavItemButton__content">
      {buttonIcon}

      <EuiInnerText>
        {(ref, innerText) => (
          <span
            ref={ref}
            title={truncate ? innerText : undefined}
            className={classNames('euiSideNavItemButton__label', {
              'euiSideNavItemButton__label--truncated': truncate,
            })}>
            {children}
          </span>
        )}
      </EuiInnerText>

      {caret}
    </span>
  );

  const renderItemProps: _EuiSideNavItemButtonProps = {
    href,
    rel,
    target,
    onClick: childrenOnly ? toggleItemOpen : onClick,
    className: buttonClasses,
    children: buttonContent,
  };
  return (
    <div className={classes}>
      <RenderItem {...renderItemProps} {...rest} />
      {childItems}
    </div>
  );
}
