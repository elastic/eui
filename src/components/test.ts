export function add(x: number, y: number) {
  return x + y;
}

add(5, 4);
