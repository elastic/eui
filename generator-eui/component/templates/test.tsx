/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { render } from '../../test/rtl';
import { requiredProps } from '../../test/required_props';

import { <%= componentName %> } from './<%= fileName %>';

describe('<%= componentName %>', () => {
  it('renders', () => {
    const { container } = render(
      <<%= componentName %> {...requiredProps} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
