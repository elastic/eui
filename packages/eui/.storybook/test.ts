/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { queries, within, waitFor, fireEvent, expect } from '@storybook/test';
import * as dataTestSubjQueries from '../src/test/rtl/data_test_subj_queries';

/**
 * Custom Storybook within util with EUI query helpers
 * + additional chained async/waitFor component utils
 *
 * @see https://storybook.js.org/docs/writing-stories/play-function#writing-stories-with-the-play-function
 * @see https://testing-library.com/docs/dom-testing-library/api-within/
 */
const customWithin = (canvasElement: HTMLElement) => {
  const canvas = within<typeof queries & typeof dataTestSubjQueries>(
    canvasElement,
    { ...queries, ...dataTestSubjQueries }
  );

  return {
    ...canvas,

    /**
     * 1. Loki doesn't like userEvent, only fireEvent
     * 2. Storybook fires fireEvents too early (esp. on page load), so we add a waitFor
     */
    waitForAndClick: async (testSubject: string) => {
      await waitFor(() =>
        expect(canvas.getByTestSubject(testSubject)).toBeInTheDocument()
      );
      await fireEvent.click(canvas.getByTestSubject(testSubject));
    },

    waitForEuiPopoverVisible: async () =>
      await waitFor(() =>
        expect(canvasElement.querySelector('[data-popover-open]')).toBeVisible()
      ),
    waitForEuiPopoverHidden: async () =>
      await waitFor(() =>
        expect(
          canvasElement.querySelector('[data-popover-panel]')
        ).not.toBeInTheDocument()
      ),
  };
};

export { customWithin as within };
