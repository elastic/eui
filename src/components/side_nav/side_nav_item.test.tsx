/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { EuiSideNavItem } from './side_nav_item';

describe('EuiSideNavItem', () => {
  test('is rendered', () => {
    const component = render(
      <EuiSideNavItem>
        <button {...requiredProps} />
      </EuiSideNavItem>
    );

    expect(component).toMatchSnapshot();
  });

  test("preserves child's classes", () => {
    const component = render(
      <EuiSideNavItem>
        <button className="test" />
      </EuiSideNavItem>
    );

    expect(component).toMatchSnapshot();
  });

  test('can have truncation turned off', () => {
    const component = render(
      <EuiSideNavItem truncate={false}>Children</EuiSideNavItem>
    );

    expect(component).toMatchSnapshot();
  });

  test('can be emphasized', () => {
    const component = render(
      <EuiSideNavItem emphasize>Children</EuiSideNavItem>
    );

    expect(component).toMatchSnapshot();
  });

  test('can be disabled', () => {
    const component = render(
      <EuiSideNavItem disabled>Children</EuiSideNavItem>
    );

    expect(component).toMatchSnapshot();
  });

  describe('isSelected', () => {
    test('defaults to false', () => {
      const component = render(
        <EuiSideNavItem>
          <button />
        </EuiSideNavItem>
      );

      expect(component).toMatchSnapshot();
    });

    test('is rendered when specified as true', () => {
      const component = render(
        <EuiSideNavItem isSelected>
          <button />
        </EuiSideNavItem>
      );

      expect(component).toMatchSnapshot();
    });
  });

  describe('href', () => {
    test('is rendered', () => {
      const component = render(
        <EuiSideNavItem href="#">
          <button />
        </EuiSideNavItem>
      );

      expect(component).toMatchSnapshot();
    });

    test('is rendered with rel', () => {
      const component = render(
        <EuiSideNavItem href="#" rel="noopener">
          <button />
        </EuiSideNavItem>
      );

      expect(component).toMatchSnapshot();
    });
  });
});
