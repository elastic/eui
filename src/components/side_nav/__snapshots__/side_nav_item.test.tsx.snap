// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiSideNavItem can be disabled 1`] = `
<div
  class="euiSideNavItem euiSideNavItem--root"
>
  <button
    class="euiSideNavItemButton"
    disabled=""
    type="button"
  >
    <span
      class="euiSideNavItemButton__content"
    >
      <span
        class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
      >
        Children
      </span>
    </span>
  </button>
</div>
`;

exports[`EuiSideNavItem can be emphasized 1`] = `
<div
  class="euiSideNavItem euiSideNavItem--root euiSideNavItem--emphasized"
>
  <div
    class="euiSideNavItemButton"
  >
    <span
      class="euiSideNavItemButton__content"
    >
      <span
        class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
      >
        Children
      </span>
    </span>
  </div>
</div>
`;

exports[`EuiSideNavItem can have truncation turned off 1`] = `
<div
  class="euiSideNavItem euiSideNavItem--root"
>
  <div
    class="euiSideNavItemButton"
  >
    <span
      class="euiSideNavItemButton__content"
    >
      <span
        class="euiSideNavItemButton__label"
      >
        Children
      </span>
    </span>
  </div>
</div>
`;

exports[`EuiSideNavItem href is rendered 1`] = `
<div
  class="euiSideNavItem euiSideNavItem--root"
>
  <a
    class="euiSideNavItemButton euiSideNavItemButton--isClickable"
    href="#"
    rel="noreferrer"
  >
    <span
      class="euiSideNavItemButton__content"
    >
      <span
        class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
      >
        <button />
      </span>
    </span>
  </a>
</div>
`;

exports[`EuiSideNavItem href is rendered with rel 1`] = `
<div
  class="euiSideNavItem euiSideNavItem--root"
>
  <a
    class="euiSideNavItemButton euiSideNavItemButton--isClickable"
    href="#"
    rel="noopener noreferrer"
  >
    <span
      class="euiSideNavItemButton__content"
    >
      <span
        class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
      >
        <button />
      </span>
    </span>
  </a>
</div>
`;

exports[`EuiSideNavItem is rendered 1`] = `
<div
  class="euiSideNavItem euiSideNavItem--root"
>
  <div
    class="euiSideNavItemButton"
  >
    <span
      class="euiSideNavItemButton__content"
    >
      <span
        class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
      >
        <button
          aria-label="aria-label"
          class="testClass1 testClass2"
          data-test-subj="test subject string"
        />
      </span>
    </span>
  </div>
</div>
`;

exports[`EuiSideNavItem isSelected defaults to false 1`] = `
<div
  class="euiSideNavItem euiSideNavItem--root"
>
  <div
    class="euiSideNavItemButton"
  >
    <span
      class="euiSideNavItemButton__content"
    >
      <span
        class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
      >
        <button />
      </span>
    </span>
  </div>
</div>
`;

exports[`EuiSideNavItem isSelected is rendered when specified as true 1`] = `
<div
  class="euiSideNavItem euiSideNavItem--root"
>
  <div
    class="euiSideNavItemButton euiSideNavItemButton-isSelected"
  >
    <span
      class="euiSideNavItemButton__content"
    >
      <span
        class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
      >
        <button />
      </span>
    </span>
  </div>
</div>
`;

exports[`EuiSideNavItem preserves child's classes 1`] = `
<div
  class="euiSideNavItem euiSideNavItem--root"
>
  <div
    class="euiSideNavItemButton"
  >
    <span
      class="euiSideNavItemButton__content"
    >
      <span
        class="euiSideNavItemButton__label euiSideNavItemButton__label--truncated"
      >
        <button
          class="test"
        />
      </span>
    </span>
  </div>
</div>
`;
